torch/csrc/jit/interpreter.cpp

```diff
@@ -1,42 +1,208 @@
+#include "Python.h"
 #include "interpreter.h"
 #include "torch/csrc/jit/ir.h"
 #include "torch/csrc/autograd/profiler.h"
 #include "torch/csrc/jit/generated/aten_dispatch.h"
+#include "Python.h"
+#include "pybind11/pybind11.h"
+#include "torch/csrc/utils/auto_gil.h"
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/autograd/python_variable.h"
+#include "torch/csrc/autograd/python_engine.h"
+#include "torch/csrc/autograd/functions/special.h"
 #ifdef WITH_CUDA
 #include "torch/csrc/jit/fusion_compiler.h"
 #endif
 
+namespace py = pybind11;
+
 namespace torch { namespace jit {
 
+struct DummyFunction : autograd::Function {
+  DummyFunction() {
+    num_inputs = 0;
+    is_executable = true;
+  }
+  virtual autograd::variable_list apply(const autograd::variable_list& inputs) override {
+    throw std::logic_error("DummyFunction::apply() called, but it should be blocked by a callback returning false");
+  }
+};
+
+struct Handle : at::RefCounted {
+  std::shared_ptr<DummyFunction> forward_inputs;
+  autograd::function_list forward_outputs;
+};
+
+at::Tensor unsafeTensorBorrow(at::RefCounted * rc) {
+  return at::Tensor(static_cast<at::TensorImpl*>(rc), true);
+}
+
+struct HandleBuilder {
+  HandleBuilder(bool requires_handle) {
+    if(requires_handle) {
+      handle = new Handle();
+      handle->forward_inputs = std::make_shared<DummyFunction>();
+    }
+  }
+  autograd::Variable addInput(at::RefCounted* input, const VariableFlags & flags_) {
+    if(handle && flags_.requires_grad) {
+      autograd::VarFlags flags = {flags_.requires_grad, flags_.is_volatile};
+      return autograd::make_variable(
+        unsafeTensorBorrow(input),
+        flags,
+        handle->forward_inputs->num_inputs++,
+        handle->forward_inputs);
+    } else {
+      autograd::VarFlags flags = {false, false};
+      return autograd::make_variable(unsafeTensorBorrow(input), flags);
+    }
```

Comment:
You're not handling volatile (which is not strictly necessary for correctness, but might be nice to have). Can you at least leave a note saying that it's on purpose or that it's a todo?